soroban-env-host/src/host.rs

```diff
@@ -2918,6 +2916,218 @@ impl VmCallerEnv for Host {
+            return Err(self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InvalidInput,
+                "g1_msm: invalid input vector lengths",
+                &[p_len.to_val(), s_len.to_val()],
+            ));
+        }
+        let points = self.g1_vec_from_vecobj(vp)?;
+        let scalars = self.scalar_vec_from_vecobj(vs)?;
+        let res = self.g1_msm_internal(&points, &scalars)?;
+        self.g1_projective_serialize_uncompressed(res)
+    }
+
+    fn bls12_381_map_fp_to_g1(
+        &self,
+        _vmcaller: &mut VmCaller<Host>,
+        fp: BytesObject,
+    ) -> Result<BytesObject, HostError> {
+        let fp = self.fp_deserialize_from_bytesobj(fp)?;
+        let g1 = self.map_fp_to_g1_internal(fp)?;
+        self.g1_affine_serialize_uncompressed(g1)
+    }
+
+    fn bls12_381_hash_to_g1(
+        &self,
+        _vmcaller: &mut VmCaller<Host>,
+        mo: BytesObject,
+        dst: BytesObject,
+    ) -> Result<BytesObject, HostError> {
+        let g1 = self.visit_obj(mo, |msg: &ScBytes| {
+            self.visit_obj(dst, |dst: &ScBytes| {
+                let dst_len = dst.len();
+                if dst_len == 0 || dst_len > 255 {
+                    return Err(self.err(
+                        ScErrorType::Crypto,
+                        ScErrorCode::InvalidInput,
+                        format!(
+                            "hash_to_g1: invalid input dst length {dst_len}, must be > 0 and < 256"
+                        )
+                        .as_str(),
+                        &[],
+                    ));
+                }
+                self.hash_to_g1_internal(dst.as_slice(), msg.as_slice())
+            })
+        })?;
+        self.g1_affine_serialize_uncompressed(g1)
+    }
+
+    fn bls12_381_g2_add(
+        &self,
+        _vmcaller: &mut VmCaller<Host>,
+        p0: BytesObject,
+        p1: BytesObject,
+    ) -> Result<BytesObject, HostError> {
+        let p0 = self.g2_affine_deserialize_from_bytesobj(p0)?;
+        let p1 = self.g2_affine_deserialize_from_bytesobj(p1)?;
+        let res = self.g2_add_internal(p0, p1)?;
+        self.g2_projective_serialize_uncompressed(res)
+    }
+
+    fn bls12_381_g2_mul(
+        &self,
+        _vmcaller: &mut VmCaller<Host>,
+        p0: BytesObject,
+        scalar_le_bytes: U256Val,
+    ) -> Result<BytesObject, HostError> {
+        let p0 = self.g2_affine_deserialize_from_bytesobj(p0)?;
+        let scalar = self.fr_from_u256val(scalar_le_bytes)?;
+        let res = self.g2_mul_internal(p0, scalar)?;
+        self.g2_projective_serialize_uncompressed(res)
+    }
+
+    fn bls12_381_g2_msm(
+        &self,
+        vmcaller: &mut VmCaller<Host>,
+        vp: VecObject,
+        vs: VecObject,
+    ) -> Result<BytesObject, HostError> {
+        let p_len = self.vec_len(vmcaller, vp)?;
+        let s_len = self.vec_len(vmcaller, vs)?;
+        if u32::from(p_len) != u32::from(s_len) || u32::from(p_len) == 0 {
```

Comment:
yeah I want to make sure to check these up front and fail as early as possible. 
The internal routines `g1_msm` performs an additional check just for extra safety, it was really meant as a `debug_assert` but I know we don't like those.

Reply:
>I want to make sure to check these up front and fail as early as possible.

I'm not sure there is any benefit to failing earlier. The code would be cleaner if every function checked its input preconditions and returned an error with a readable diagnostic. Here we have some code that doesn't quite belong here on one hand, and we also have a 'redundant' check in the library on the other hand. Why not leave just the more important check (in the library)?

This pattern is repeated in several other host methods and I think it applies there as well. Basically we probably want to be defensive in the function implementations to make sure they never get bad inputs, and I'm not sure it makes sense to also have similar checks at a higher level.

Reply:
The benefit is a performance optimization. If we know the length check doesn't match, we fail immediately instead of doing a bunch of redundant work, this speeds up the transaction by an amount. 
Previously the difference probably don't matter, but for these functions the vector parsing can be fairly expensive, 
 a few millions cpu instructions. 
That'll make transaction closing, replay go faster if there are many such failures. I think that's an optimization worth taking?

Reply:
(I get the redundant condition check and cleanness aspect though)

Reply:
In general I don't think it's worth optimizing the failure mode. If transaction was simulated with success result, it still will take up ledger space (not to mention that failure scenario is highly unlikely for this function set - why would someone send transactions with invalid signatures that are likely guaranteed to fail).

But in this particular case there is even no need to debate whether this is slower as the length check happens before we actually get to more expensive deserialization parts (at least it should happen before deserialization for the sake of correctness).

Reply:
removed the higher level check and improved the message on the actual check

Reply:
Thanks!